src/catchup/CatchupManagerImpl.cpp

```diff
@@ -93,4 +217,80 @@ CatchupManagerImpl::logAndUpdateCatchupStatus(bool contiguous)
 {
     logAndUpdateCatchupStatus(contiguous, getStatus());
 }
+
+bool
+CatchupManagerImpl::hasBufferedLedger() const
+{
+    return !mSyncingLedgers.empty();
+}
+
+LedgerCloseData const&
+CatchupManagerImpl::getBufferedLedger() const
+{
+    if (!hasBufferedLedger())
+    {
+        throw std::runtime_error(
+            "getBufferedLedger called when mSyncingLedgers is empty!");
+    }
+
+    return mSyncingLedgers.front();
+}
+
+void
+CatchupManagerImpl::popBufferedLedger()
+{
+    if (!hasBufferedLedger())
+    {
+        throw std::runtime_error(
+            "popBufferedLedger called when mSyncingLedgers is empty!");
+    }
+
+    mSyncingLedgers.pop_front();
+    mSyncingLedgersSize.set_count(mSyncingLedgers.size());
+}
+
+void
+CatchupManagerImpl::syncMetrics()
+{
+    mSyncingLedgersSize.set_count(mSyncingLedgers.size());
+}
+
+void
+CatchupManagerImpl::reset()
+{
+    mCatchupWork.reset();
```

Comment:
I don't think `reset` should work like this as it's unsafe:
it should perform this reset only if `mCatchupWork` is done.
This avoids a potential race with `LedgerManager` resetting `mCatchupWork` and right after going back into "catchup" state (now we'd potentially have multiple `mCatchupWork` running)

Reply:
this `reset` function is also the one that should call `logAndUpdateCatchupStatus` to clear the catchup status

Reply:
Made `reset` private and only reset `mCatchupWork` when it's done.